### ISSUE TITLE
Feature request: stroke alignment inside/outside

### ISSUE THREAD

**Issue by craigsnyders** [openlayers/openlayers#7932]
I would like to request the ability to position the stroke (of a polygon) as `inside` or `outside` its path. With this I could:

- create interesting custom feature interactions that need `inside` or `outside` stroke to look good
- easily distinguish borders between shapes where the exact border is important, for example land parcels which are individually coloured. Current OL implementation would have 'overlapping' stroke borders, whereas I would like the edges to be distinguishable and crisp.

E.G. `alignment: "inside"`
 ![image](https://user-images.githubusercontent.com/3810017/36975383-ce11d22c-2071-11e8-8347-8a80b8590df3.png)

Leaflet have [considered this feature](https://github.com/Leaflet/Leaflet/issues/3665) but I don't think they have implemented anything.
Mapbox GL JS (WebGL, not canvas) supports this via [line-offset](https://www.mapbox.com/mapbox-gl-js/style-spec/#paint-line-line-offset).

**Comment by fredj** [comment]
We will also need an inside stroke alignement for a project; I've started to work on this, see:
https://github.com/openlayers/openlayers/compare/master...fredj:stroke_alignment

**Comment by craigsnyders** [comment]
@fredj nice work on your branch. Are you planning to do any more work on it? I don't know the canvas API that well but I could take a crack too

**Comment by jokd** [comment]
@fredj What is left to do on this? Would be a great addition and perhaps I could check it out.

**Comment by martinmcneela** [comment]
Is this something you can look at reopening?

**Comment by jahow** [comment]
This would be covered by https://github.com/openlayers/openlayers/issues/9377

**Comment by martinmcneela** [comment]
@jahow I've had a look at #9377 and at the code for v6.0.0-beta.11 and this doesn't seem to be covered - with #9377 being closed-off already.

**Comment by jahow** [comment]
Sorry if that wasn't clear, #9377 is simply a proposal for a more general implementation of an offset style parameter. It was closed for inactivity.

**Comment by ezriYerushalmi** [comment]
any new about this?

**Comment by elliotlarson** [comment]
Yeah, it seems like this was closed without addressing the issue.  It would be nice to have the ability to set the stroke alignment property.  Can we re-open this issue?

**Comment by warm-coolguy** [comment]
Plus for re-opening. This has wide applications (e.g. municipal boundaries or parcels) and the linked issues do not directly address this, but have a related topic; just not quite the same.

**Comment by swimaf** [comment]
While waiting for the offset functionality, I started developing a function that would add an inside and/or outside border. I use the turf library

![Capture d'écran_20221205_191623](https://user-images.githubusercontent.com/5155123/205712114-f31d8071-f2a0-4e55-a9a9-a424639d0f0b.png)

  ```
import {Circle as CircleStyle, Fill, RegularShape, Stroke, Style, Text} from 'ol/style';
import {LineString, Polygon, Point, Circle} from 'ol/geom';
import * as math from "mathjs";
import unkinkPolygon from "@turf/unkink-polygon";
import {polygon} from "@turf/helpers";
import booleanClockwise from "@turf/boolean-clockwise";


setBorderInner(geometry, resolution) {
        let result = unkinkPolygon(polygon([geometry.getLinearRing(0).getCoordinates()]));
        let styles = []
        result.features.forEach((polygon) => {
            styles.push(...this._addBorderInner(new LineString(polygon.geometry.coordinates[0]), resolution))
        })
        return styles
    } 
```
```
    _addBorderInner(geometry, resolution) {
        var styles = []
        let clockWise = booleanClockwise(geometry.getCoordinates())

        let colors = ['rgba(37,147,7,0.4)', 'transparent'];
        if(clockWise) {
            colors = colors.reverse()
        }
        let width = 8
        for (let line = 0; line < colors.length; line++) {
            let dist = width / (resolution/10) * (line - (colors.length-1)/2);
            let coords = [];
            let counter = 0;
            let coordinates = geometry.getCoordinates()
            let from = coordinates[coordinates.length-2]
            coordinates.forEach(to => {
                let angle = Math.atan2(to[1] - from[1], to[0] - from[0]);
                let newFrom = [
                    Math.sin(angle) * dist + from[0],
                    -Math.cos(angle) * dist + from[1]
                ];
                var newTo = [
                    Math.sin(angle) * dist + to[0],
                    -Math.cos(angle) * dist + to[1]
                ];
                coords.push(newFrom);
                coords.push(newTo);
                if (coords.length > 2) {
                    let intersection = math.intersect(coords[counter], coords[counter+1], coords[counter+2], coords[counter+3]);
                    coords[counter+1] = (intersection) ? intersection : coords[counter+1];
                    coords[counter+2] = (intersection) ? intersection : coords[counter+2];
                    counter += 2;
                }

                from = to;
            });
            coordinates.shift()
            coords.shift()
            coords.shift()
            styles.push(this._addStyleBorder(coordinates, coords, colors[line]))

        }
        return styles
    }
```
```
    _addStyleBorder(coordinates, coords, color) {
        var styles = []
        for (let i = 0; i < coordinates.length; i++) {
            let c = ((i+1) % coordinates.length)
            styles.push( new Style({
                geometry: new Polygon([[coordinates[i], coordinates[c], coords[c*2], coords[i*2], coordinates[i]]]),
                fill: new Fill({
                    color: color,
                })
            }))
        }
        return styles
    }`
```

**Comment by dbauszus-glx** [comment]
Has anybody had another look at stroke style allignment to be inside/outside of a polygon?

**Comment by jahow** [comment]
This is something you can now do with the WebGL vector layer renderer with the `stroke-offset` style property.

**Comment by dbauszus-glx** [comment]
@jahow Is there a hosted example for this?

**Comment by jahow** [comment]
You can play with it here: https://openlayers.org/en/latest/examples/webgl-draw-line.html

**Comment by dbauszus-glx** [comment]
@ahocevar Thanks Andreas. I can imagine this being a tricky task. I assume the only way this can happen is by clipping the canvas path for inside lines or draw the fill on top for outside lines. Writing this I don't think there are many use cases for outside drawn paths. It is only the inside drawn path which is useful like shown in the orginal post image. Instead of defining alignment I would be happy if there is a clip [outside] flag in the stroke properties which would only apply in the canvas PolygonBuilder.

Is this the right ticket for this task?

Something I would like to have another look at if I am going to organise a code sprint this year.

**Comment by ahocevar** [comment]
@dbauszus-glx Canvas clipping would only work for the special cases with stroke inside or outside, but starting at the polygon boundary. My goal for a stroke-offset property was to actually calculate offset geometries. We could make good use of that also for offsetting text along lines. The tricky part here is a reliable algorithm for offsetting lines.
